### ISSUE TITLE
Socket hangs up in some cases

### ISSUE THREAD

**Issue by idy** [go-mgo/mgo#502]
I thinks the lock before socket.conn.write cause the problem: https://github.com/globalsign/mgo/blob/master/socket.go#L522

The unlock of this mutx is after the write opreation: https://github.com/globalsign/mgo/blob/master/socket.go#L561

Which caused the socket cannot read while writing.

**Comment by hexadecy** [comment]
@idy probably because of that:
> Is calling read and write functions from different goroutines on the
> same net.TCPConn object safe?

> if you call read from one and write from the other, yes.
> if you call read from two or write from two, then they
> will get interleaved.
> russ

I think we found a issue https://github.com/globalsign/mgo/issues/60 with https://github.com/globalsign/mgo/pull/52

In our case it is possible to have at least 2-3 goroutines for the same socket, as we use session.clone for our http handlers.
In the past, I resolved a similar issue by switching to session.copy for all our cronjob like goroutines.
I will keep searching to reproduce it, as right now it takes at least 2 days to have it. Maybe it's just Windows related, I will also check Linux.

**Comment by idy** [comment]
@hexadecy  Your quote is a post before year 2011, and it's not a official one. According to the official documentation, the net.TCPConn is thread safe. If some version go doesn't follow the behavious described in doc, it's a issue of go not the library.

**Comment by hexadecy** [comment]
@idy, yes you are right, but I see more timeout when the socket read and write at the same time with https://github.com/globalsign/mgo/commit/5be15cc6a44937b6937a9487e1633f8a5fb25e4a. This https://github.com/globalsign/mgo/pull/69 will surely help, I have not try it yet. Do you use **session.Copy**?
